cmds/bootcm4.c

```diff
@@ -0,0 +1,192 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system loader
+ *
+ * boot CM4 CPU i.MX RT117x
+ *
+ * Copyright 2022-2023 Phoenix Systems
+ * Author: Gerard Swiderski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "cmd.h"
+
+#include <hal/hal.h>
+#include <phfs/phfs.h>
+#include <lib/lib.h>
+
+
+#define CM4_BOOT_MEMSIZE 0x00040000u
+#define CM4_BOOT_ADDRESS 0x20200000u
+
+
+static struct {
+	unsigned int isImgLoaded;
+	addr_t cm4VtorOffset;
+} common = { 0 };
+
+
+static void cmd_bootcm4Info(void)
+{
+	lib_printf("boots CM4 core image, use <-h> to see usage");
+}
+
+
+static void print_usage(const char *name)
+{
+	lib_printf(
+		"Usage: %s <options> [<device> <image>]\n"
+		"\t-b  boot CM4 core\n"
+		"\t-l  load binary image using phfs <device> and <image>\n"
+		"\t-V  optional offset to vectors table (default=0)\n",
+		name);
+}
+
+
+static ssize_t loadImagePhfs(const char *phfsDev, const char *phfsFile)
+{
+	u8 buff[SIZE_MSG_BUFF];
+	handler_t h;
+	addr_t offs = 0u;
+	addr_t addr = CM4_BOOT_ADDRESS;
+
+	ssize_t res = phfs_open(phfsDev, phfsFile, 0, &h);
+
+	if (res < 0) {
+		log_error("Can't open '%s' on '%s'", phfsFile, phfsDev);
+		return res;
+	}
+
+	do {
+		res = phfs_read(h, offs, buff, sizeof(buff));
+		if (res < 0) {
+			log_error("Can't read '%s' from '%s'", phfsFile, phfsDev);
+			break;
+		}
+
+		offs += res;
+		if (offs > CM4_BOOT_MEMSIZE) {
+			log_error("Image '%s' too big", phfsFile);
+			res = -ENOSPC;
+			break;
+		}
+
+		hal_memcpy((void *)addr, buff, res);
+		addr += res;
+	} while (res > 0);
+
+	phfs_close(h);
+
+	return res;
+}
+
+
+static int cmd_bootcm4(int argc, char *argv[])
+{
+	char *endptr;
+	const char *phfsDev = NULL;
+	const char *phfsFile = NULL;
+
+	int opt;
+	int optLoad = 0;
+	int optBoot = 0;
+
+	lib_printf("\n");
+
+	if (argc == 1) {
+		print_usage(argv[0]);
+		return CMD_EXIT_FAILURE;
+	}
+
+	for (;;) {
+		opt = lib_getopt(argc, argv, "bhlV:");
+		if (opt == -1) {
+			break;
+		}
+
+		switch (opt) {
+			case 'b':
+				optBoot = 1;
+				break;
+
+			case 'l':
+				optLoad = 1;
+				break;
+
+			case 'V':
```

Comment:
I would prefer lower case 'o' or 'v' here. I don't like mixing lower case & upper case options unless it's necessary.

Reply:
The lower case `v` is most often used to retrieve version information or set the verbosity level. Changed to `o`.